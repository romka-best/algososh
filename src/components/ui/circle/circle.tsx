import React from 'react';
import cn from 'classnames';

import { ElementStates } from '../../../types/element-states';

import styles from './circle.module.css';

interface CircleProps {
    state?: ElementStates;
    letter?: string;
    head?: string | React.ReactElement | null;
    index?: number;
    tail?: string | React.ReactElement | null;
    tailType?: 'string' | 'element';
    extraClass?: string;
    isSmall?: boolean;
}

export const Circle: React.FC<CircleProps> = (
    {
        state = ElementStates.Default,
        letter,
        head,
        index,
        tail,
        extraClass = '',
        isSmall,
    }
) => {
    return (
        <div className={ cn(styles.content, extraClass) }>
            <div
                className={ cn('text',
                    'text_type_input',
                    'text_color_input',
                    'mb-4',
                    styles.absolute,
                    styles.head,
                    styles[typeof head === 'string' ? 'string' : 'element'],
                )
                }
            >
                { head }
            </div>
            <div
                className={ cn(styles.circle, isSmall && styles.small, styles[state]) }
            >
                { letter && (
                    <p
                        className={ cn('text', 'text_type_circle', 'text_color_input', styles.letter) }
                    >
                        { letter }
                    </p>
                ) }
            </div>
            { index !== undefined && (
                <p
                    className={ cn('text',
                        'text_type_input',
                        'text_color_input',
                        'mt-4',
                        styles.absolute,
                        styles.index,
                    ) }
                >
                    { index.toString() }
                </p>
            ) }
            <div
                className={ cn(
                    'text',
                    'text_type_input',
                    'text_color_input',
                    'mt-4',
                    styles.absolute,
                    styles.tail,
                    index?.toString() ? styles.tail60 : styles.tail30,
                    styles[typeof tail === 'string' ? 'string' : 'element'],
                ) }
            >
                { tail }
            </div>
        </div>
    );
};
