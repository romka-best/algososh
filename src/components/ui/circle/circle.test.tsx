import React from 'react';
import renderer from 'react-test-renderer';

import { ElementStates } from '../../../types/element-states';

import { Circle } from './circle';

describe('Circle', () => {
    describe('Отображение', () => {
        it('Без буквы', () => {
            const circle = renderer
                .create(
                    <Circle/>
                )
                .toJSON();

            expect(circle).toMatchSnapshot();
        });

        it('С буквой', () => {
            const circle = renderer
                .create(
                    <Circle
                        letter="F"
                    />
                )
                .toJSON();

            expect(circle).toMatchSnapshot();
        });

        describe('С head', () => {
            it('Строка', () => {
                const circle = renderer
                    .create(
                        <Circle
                            head="top"
                        />
                    )
                    .toJSON();

                expect(circle).toMatchSnapshot();
            });

            it('React-элемент', () => {
                const circle = renderer
                    .create(
                        <Circle
                            head={
                                (
                                    <div>
                                        Здесь будет голова!
                                    </div>
                                )
                            }
                        />
                    )
                    .toJSON();

                expect(circle).toMatchSnapshot();
            });
        });

        describe('С tail', () => {
            it('Строка', () => {
                const circle = renderer
                    .create(
                        <Circle
                            tail="bottom"
                        />
                    )
                    .toJSON();

                expect(circle).toMatchSnapshot();
            });

            it('React-элемент', () => {
                const circle = renderer
                    .create(
                        <Circle
                            tail={
                                (
                                    <div>
                                        Здесь будет хвост!
                                    </div>
                                )
                            }
                        />
                    )
                    .toJSON();

                expect(circle).toMatchSnapshot();
            });
        });

        it('С index', () => {
            const circle = renderer
                .create(
                    <Circle
                        index={ 0 }
                    />
                )
                .toJSON();

            expect(circle).toMatchSnapshot();
        });

        describe('Разные размеры', () => {
            it('Дефолтный', () => {
                const circle = renderer
                    .create(
                        <Circle
                            isSmall={ false }
                        />
                    )
                    .toJSON();

                expect(circle).toMatchSnapshot();
            });

            it('Маленький', () => {
                const circle = renderer
                    .create(
                        <Circle
                            isSmall={ true }
                        />
                    )
                    .toJSON();

                expect(circle).toMatchSnapshot();
            });
        });

        describe('Разные состояния', () => {
            it('Default', () => {
                const circle = renderer
                    .create(
                        <Circle
                            state={ ElementStates.Default }
                        />
                    )
                    .toJSON();

                expect(circle).toMatchSnapshot();
            });

            it('Changing', () => {
                const circle = renderer
                    .create(
                        <Circle
                            state={ ElementStates.Changing }
                        />
                    )
                    .toJSON();

                expect(circle).toMatchSnapshot();
            });

            it('Modified', () => {
                const circle = renderer
                    .create(
                        <Circle
                            state={ ElementStates.Modified }
                        />
                    )
                    .toJSON();

                expect(circle).toMatchSnapshot();
            });
        });
    });
});
