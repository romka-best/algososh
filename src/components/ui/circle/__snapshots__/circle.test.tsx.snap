// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle Отображение Без буквы 1`] = `
<div
  className="content"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle default"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail tail30 element"
  />
</div>
`;

exports[`Circle Отображение Разные размеры Дефолтный 1`] = `
<div
  className="content"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle default"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail tail30 element"
  />
</div>
`;

exports[`Circle Отображение Разные размеры Маленький 1`] = `
<div
  className="content"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle small default"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail tail30 element"
  />
</div>
`;

exports[`Circle Отображение Разные состояния Changing 1`] = `
<div
  className="content"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle changing"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail tail30 element"
  />
</div>
`;

exports[`Circle Отображение Разные состояния Default 1`] = `
<div
  className="content"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle default"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail tail30 element"
  />
</div>
`;

exports[`Circle Отображение Разные состояния Modified 1`] = `
<div
  className="content"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle modified"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail tail30 element"
  />
</div>
`;

exports[`Circle Отображение С head React-элемент 1`] = `
<div
  className="content"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  >
    <div>
      Здесь будет голова!
    </div>
  </div>
  <div
    className="circle default"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail tail30 element"
  />
</div>
`;

exports[`Circle Отображение С head Строка 1`] = `
<div
  className="content"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head string"
  >
    top
  </div>
  <div
    className="circle default"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail tail30 element"
  />
</div>
`;

exports[`Circle Отображение С index 1`] = `
<div
  className="content"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle default"
  />
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  >
    0
  </p>
  <div
    className="text text_type_input text_color_input mt-4 absolute tail tail60 element"
  />
</div>
`;

exports[`Circle Отображение С tail React-элемент 1`] = `
<div
  className="content"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle default"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail tail30 element"
  >
    <div>
      Здесь будет хвост!
    </div>
  </div>
</div>
`;

exports[`Circle Отображение С tail Строка 1`] = `
<div
  className="content"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle default"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail tail30 string"
  >
    bottom
  </div>
</div>
`;

exports[`Circle Отображение С буквой 1`] = `
<div
  className="content"
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      F
    </p>
  </div>
  <div
    className="text text_type_input text_color_input mt-4 absolute tail tail30 element"
  />
</div>
`;
